extract-locales.js

```diff
@@ -0,0 +1,79 @@
+/**
+ *    Copyright (c) ppy Pty Ltd <contact@ppy.sh>.
+ *
+ *    This file is part of osu!web. osu!web is distributed with the hope of
+ *    attracting more community contributions to the core ecosystem of osu!.
+ *
+ *    osu!web is free software: you can redistribute it and/or modify
+ *    it under the terms of the Affero GNU General Public License version 3
+ *    as published by the Free Software Foundation.
+ *
+ *    osu!web is distributed WITHOUT ANY WARRANTY; without even the implied
+ *    warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ *    See the GNU Affero General Public License for more details.
+ *
+ *    You should have received a copy of the GNU Affero General Public License
+ *    along with osu!web.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+'use strict';
+
+const fs = require('fs');
+const path = require('path');
+const mkdirp = require('mkdirp');
+
+const buildPath = path.resolve(__dirname, 'resources/assets/build');
+const localesPath = path.resolve(buildPath, 'locales');
+const messagesPath = path.resolve(buildPath, 'messages.json');
+
+function extractLanguages() {
+  console.log('Extracting localizations...')
+  mkdirp.sync(localesPath);
+
+  const messages = getAllMesssages();
+
+  const languages = new Map();
+  for (const key in messages) {
+    const index = key.indexOf('.');
+    const language = key.substring(0, index);
+    if (!languages.has(language)) {
+      languages.set(language, {});
+    }
+    languages.get(language)[key] = messages[key];
+  }
+
+  return languages;
+}
+
+function getAllMesssages() {
+  const content = fs.readFileSync(messagesPath);
+
+  return JSON.parse(content);
+}
+
+function runLangJs()
+{
+  const { spawnSync } = require('child_process');
+  spawnSync('php', ['artisan', 'lang:js', '--json', messagesPath], { stdio: 'inherit' });
```

Comment:
looks like docker needs updating to handle this as currently `yarn` container doesn't have `php` and vice versa. It's already not quite working smoothly as one needs to be re-run (`yarn` container, to build correct js) after the other (`php` container, to generate the `messages.js`) but at least it works in the end.

I think the separation should be `php-fpm` and `assets-builder` instead of `php` and `yarn`. Then maybe even make language (and route) js generation as part of webpack.

Reply:
Another problem to be solved another time ™️